broker/fragment/store_azure.go

```diff
@@ -29,30 +35,63 @@ type azureBackend struct {
 	endpoint    string
 	accountName string
 	client      pipeline.Pipeline
-	credentials *azblob.SharedKeyCredential
+	svcClient   service.Client
 	clientMu    sync.Mutex
+	udc         *service.UserDelegationCredential
+	udc_exp     *time.Time
 }
 
 func (a *azureBackend) Provider() string {
 	return "azure"
 }
 
+// Cache UserDelegationCredentials and refresh them when needed
+func (a *azureBackend) GetUserDelegationCredential() (*service.UserDelegationCredential, error) {
+	if a.udc == nil || (a.udc_exp != nil && a.udc_exp.After(time.Now().Add(time.Minute*10))) {
+		expTime := time.Now().Add(time.Hour * 24)
+		info := service.KeyInfo{
+			Start:  to.Ptr(time.Now().Add(time.Second * -10).UTC().Format(sas.TimeFormat)),
+			Expiry: to.Ptr(expTime.UTC().Format(sas.TimeFormat)),
+		}
+
+		udc, err := a.svcClient.GetUserDelegationCredential(context.Background(), info, nil)
+
+		if err != nil {
+			return nil, err
+		}
+
+		a.udc = udc
+		a.udc_exp = &expTime
+	}
+
+	return a.udc, nil
+}
+
+// https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/service/examples_test.go#L285
 func (a *azureBackend) SignGet(ep *url.URL, fragment pb.Fragment, d time.Duration) (string, error) {
 	cfg, _, err := a.azureClient(ep)
 	if err != nil {
 		return "", err
 	}
 	blobName := cfg.rewritePath(cfg.prefix, fragment.ContentPath())
-	sasQueryParams, err := azblob.BlobSASSignatureValues{
-		Protocol:      azblob.SASProtocolHTTPS, // Users MUST use HTTPS (not HTTP)
+
+	udc, err := a.GetUserDelegationCredential()
+
```

Comment:
nit: extra newline

Reply:
Normally I'd just make this change since it's a nit-pick, but in light of your meta-comment above about the burden of common understanding... my understanding was that Go has a blessed formatter that is supposed to take care of all of these stylistic "tabs vs spaces" style issues, and I ran it (and it moved some stuff around and updated some formatting, yay!).

Since it didn't catch this (or the variable thing above), I assumed that it was good to go. Since that's not the case, can you point me to some resource that lays out these things, so that I can learn where I'm supposed to put newlines or not? There are clearly newlines between other bits of code, so I'm guessing the heuristic here is "no newlines between an error-producing statement and the error-checking block" which seems perfectly reasonable, just not something I would have known.

Reply:
https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

Go fmt doesn't strip newlines as Rust's does. It's better than nothing, but Go's formatter was the first kid on the block in a new era of code formatters becoming dejour, and I believe for that reason it's less aggressive, opinionated, and configurable than formatters have become.

Other good material for Go:
* https://go.dev/doc/effective_go
* https://github.com/golang/go/wiki/CodeReviewComments

The `var` thing is a Gazette and former Arbor-ism. At one point I'd put together a whole style guide, but I'm afraid that's lost to time and private repos I don't have access too, and at this point it's institutional knowledge.